Update site to use Polymer 2.0

keanulee opened the pull request:
Fixes #2242 

Staged at https://polymer-2-dot-polymer-project.appspot.com/ - please leave feedback as comments here on this PR.

TODOs:
- [x] Make elements hybrid
- [x] Compile to ES5 and make sure webcomponents-loader.js works
- [x] Transpile-to-ES5/minify/bundle demos in /1.0/samples/homepage/
  - [x] Optional: Make homepage demos use webcomponents-loader.js instead of -lite.js
    - ~Not possible since `customElements.define()` call is bundled in index.html and must be polyfilled before in index.html~
    - The 2 homepage demos are now treated as standalone projects with their own `polymer.json` and built/bundled using `polymer build`. Element definitions are in the shell HTML import, so polyfills are guaranteed to be run by the time the shell is imported.
- [x] Page navigation seems broken in Firefox/Edge/IE (does full-page reload)
- [x] [IE] Promise.catch missing in IE - see https://github.com/webcomponents/webcomponentsjs/issues/837
  - [x] [IE] `<a>` link click handler doesn't stop page navigation (version selector now works though) https://github.com/PolymerElements/iron-location/issues/93
  - [x] Waiting for the above 2 patches to be merged and tagged
- [x] [IE] Issue with template polyfill (t.content.querySelectorAll missing) on homepage demos
- [x] Port summit 2015 and 2016 sites, or check in Polymer 1 deps in those dirs
- [x] Service worker should precache webcomponents-loader.js and runtime cache needed polyfill

keanulee commented:
The staging website (https://polymer-2-dot-polymer-project.appspot.com) is now ready for testing on all browsers, including IE 11! I'm waiting until patched dependencies are merged and tagged before merging this PR though.

notwaldorf commented:
I have been browsing and refreshing this for the last 5 minute on IE11, and it's pretty aces!

notwaldorf commented on app/2.0/samples/homepage/contact-card/index.html:
```diff
@@ -1,7 +1,7 @@
 <html lang="en">
   <head>
-    <script src="../../../../bower_components/webcomponentsjs/webcomponents-lite.min.js"></script>
+    <script src="bower_components/webcomponentsjs/webcomponents-loader.js"></script>
```
yasssssss

notwaldorf commented on app/2.0/samples/homepage/google-map/bower_components:
```diff
@@ -0,0 +1 @@
+../../../../bower_components
```
oh is this a link? smart!

notwaldorf commented on app/elements/demo-tabs.html:
```diff
@@ -93,12 +93,12 @@
 
     <div class="results">
-      <content select="iframe"></content>
+      <slot select="iframe"></slot>
```
this should be `name="iframe"`, no?

keanulee replied:
It _should_ be `<slot name="iframe">` with the corresponding iframes updated with `slot="iframe"`, but since there's no default slot before I'm just going to make this the default (and remove `select="iframe"`).

notwaldorf commented on app/elements/lazy-elements.html:
```diff
@@ -7,15 +7,15 @@
 <link rel="import" href="../bower_components/paper-icon-button/paper-icon-button.html">
 <link rel="import" href="../bower_components/paper-item/paper-item.html">
 <link rel="import" href="../bower_components/paper-listbox/paper-listbox.html">
-<link rel="import" href="../bower_components/paper-toast/paper-toast.html">
```
isn't the toast needed by the service worker code?

notwaldorf replied:
(like, why isn't it lazy anymore)

keanulee replied:
On Firefox (at least), if you navigate to a non-cached page, the toast won't be upgraded and you don't see it (and get an ugly "show is not a function" error).

notwaldorf commented on app/elements/pw-shell.html:
```diff
@@ -340,15 +341,18 @@
-        Polymer.dom(this).appendChild(document.importNode(fragment, true));
+        // Replace shell contents.
+        var shell = Polymer.dom(this);
+        while (shell.firstChild) {
```
Why is this better than `shell.innerHTML = '';`?

keanulee replied:
DOM manipulations are faster than invoking the parser, I think (even for empty strings), though the real motivation behind this change was that setting `innerHTML` was causing issues.

notwaldorf commented on bower.json:
```diff
@@ -29,13 +30,14 @@
-    "iron-selector": "2",
-    "webcomponentsjs": "^0.7.24"
+    "iron-location": "fix-ie-capture",
+    "webcomponentsjs": "master"
```
but you're going to update this to depend on real releases before you merge this, right?

notwaldorf commented on templates/head-meta.html:
```diff
@@ -14,8 +14,11 @@
-  window.Polymer = window.Polymer || {dom: 'shadow', lazyRegister: true};
-</script>
-<!-- Node: the webcomponents polyfill is lazily loaded in app.js -->
+<script>!function(e){var r=e.babelHelpers={};r.typeof="function"==typeof Symbol&&"symbol"==typeof Symbol.iterator?function(e){return typeof e}:function(e){return e&&"function"==typeof Symbol&&e.constructor===Symbol&&e!==Symbol.prototype?"symbol":typeof e},r.classCallCheck=function(e,r){if(!(e instanceof r))throw new TypeError("Cannot call a class as a function")},r.createClass=function(){function ...
```
Yum. What is this?

FluorescentHallucinogen commented:
@keanulee Could you please also update the old and/or deprecated versions of dependencies such as `app-layout#^0.0.14` and `firebase-element#^1.0.12` to actual versions?

FluorescentHallucinogen commented:
@keanulee Oh, dependenies are already updated. Please ignore my previous message. One question: why you don't use version ranges in `bower.json` for some dependenies (e.g. `paper-item#^2.0.0` instead of `paper-item#2`)?

keanulee commented:
@FluorescentHallucinogen I like `#2` because it's succinct. It equivalent to `#^2.0.0`, just personal preference 😉  (Note that `#2.2.5` is NOT equivalent to `#^2.2.5` - `2` means versions that "start with 2, where the latter is an exact match)

arthurevans commented:
Is the IE iron-location issue the last issue blocking this?

keanulee commented:
The iron-location fix is now published and I updated the staging site with the latest bower deps (https://polymer-2-dot-polymer-project.appspot.com/). I tested the latest update and everything seems to be good. @arthurevans wanna give it one last check and merge? (I believe I already addressed all @notwaldorf 's feedback.)

arthurevans commented on bower.json:
```diff
@@ -29,13 +29,9 @@
-    "promise-polyfill": "polymerlabs/promise-polyfill#^1.0.0",
-    "google-youtube": "GoogleWebComponents/google-youtube#^1.2.1",
-    "google-map": "GoogleWebComponents/google-map#^1.1.10",
+    "promise-polyfill": "polymerlabs/promise-polyfill#2",
```
I think we can drop the separate promise-polyfill now... There's a promise polyfill in the v1 WC polyfills, so WC-loader should take care of this, I think. (Only needed this to run the hybrid iron-ajax against 1.x, IIRC).

arthurevans commented:
Will pull and do a test build tomorrow to make sure I don't have any issues building locally.
